scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,385 @@
+                    for tensors in self.data_loaders_loop():
+                        if train_discriminator:
+                            latent_tensors = []
+                            for (i, (data, *_)) in enumerate(tensors):
+                                z = self.model.sample_from_posterior_z(
+                                    data, mode=i, deterministic=False
+                                )
+                                latent_tensors.append(z)
+
+                            # Train discriminator
+                            d_loss = self.loss_discriminator(
+                                [t.detach() for t in latent_tensors], True
+                            )
+                            d_loss *= self.kappa
+                            d_optimizer.zero_grad()
+                            d_loss.backward()
+                            d_optimizer.step()
+
+                            # Train generative model to fool discriminator
+                            fool_loss = self.loss_discriminator(latent_tensors, False)
+                            fool_loss *= self.kappa
+                            g_optimizer.zero_grad()
+                            fool_loss.backward()
+                            g_optimizer.step()
+
+                        # Train generative model
+                        g_loss = self.loss(tensors)
+                        g_optimizer.zero_grad()
+                        g_loss.backward()
+                        g_optimizer.step()
+
+                    if not self.on_epoch_end():
+                        break
+
+        self.model.eval()
+        self.training_time += (time.time() - begin) - self.compute_metrics_time
+        if self.frequency:
+            logger.debug(
+                "\nTraining time:  %i s. / %i epochs"
+                % (int(self.training_time), self.n_epochs)
+            )
+
+    def loss_discriminator(
+        self,
+        latent_tensors: List[torch.Tensor],
+        predict_true_class: bool = True,
+        return_details: bool = False,
+    ) -> Union[List[torch.Tensor], torch.Tensor]:
+        """
+        Compute the loss of the discriminator (either for the true labels or the fool labels)
+
+        :param latent_tensors: Tensors for each dataset of the latent space
+        :param predict_true_class: Specify if the loss aims at minimizing the accuracy or the mixing
+        :param return_details: Boolean used to inspect the loss values, return detailed loss for each dataset
+        :return: scalar loss if return_details is False, else list of scalar losses for each dataset
+        """
+        n_classes = self.n_dataset
+        losses = []
+        for i, z in enumerate(latent_tensors):
+            cls_logits = nn.LogSoftmax(dim=1)(self.discriminator(z))
+
+            if predict_true_class:
+                cls_target = torch.zeros(
+                    n_classes, dtype=torch.float32, device=z.device
+                )
+                cls_target[i] = 1.0
```

Comment:
Thanks for pointing that out, I was not that familiar with this. However for GimVI, I am not sure the adversarial training is the limiting factor of model. It will take time to benchmark with this again so I keep that in mind for future work but I won't add it in the code.